tensorcircuit/templates/ensemble.py

```diff
@@ -0,0 +1,150 @@
+"""
+Useful utilities for ensemble
+"""
+
+from typing import Any, List
+import tensorflow as tf
+import keras
+import numpy as np
+
+NDArray = Any
+kwargus = Any
+
+
+class bagging:  # A.K.A. voting
+    def __init__(self) -> None:
+        self.models: List[keras.engine.functional.Functional] = []
+        self.model_trained: List[bool] = []
+        self.count = 0
+        self.need_confidence = True  # Help in reducing numbers of get_confidence runs
+        self.permit_train = False
+
+    def append(
+        self, model: keras.engine.functional.Functional, model_trained: bool
+    ) -> None:
+        """
+        Add model to the voting method
+        """
+        self.models.append(model)
+        self.model_trained.append(model_trained)
+        self.count += 1
+
+    def __train_model(self, i: int, **kwargs: kwargus) -> None:
+        """
+        Train a model if it isn't trained already
+        """
+        if not self.model_trained[i]:
+            self.need_confidence = True
+            self.model_trained[i] = True
+            self.models[i].trainable
+            self.models[i].fit(**kwargs)
+
+    def train(self, **kwargs: kwargus) -> None:
+        """
+        Train all models in the class, **kwargs expect to receive the argus that can be directly sent to tf.fit
+        Expected to be run after finishing compile
+        """
+        if not self.permit_train:
+            # raise Exception("Needed to be compiled before training")
+            raise ValueError()
```

Comment:
the str in Exception can be taken into ValueError

Reply:
Updated